### PR TITLE
VW MQB: Longitudinal safety

### PR DESCRIPTION
**Pull request by jyoung8607**
Support for openpilot longitudinal control. I've yet to take this version on the road, but I think it's reasonably close to final.

### CODE REVIEW

**Comment by jyoung8607** [comment]
CI is going to highlight two classes of problems in `test_tx_hook_on_wrong_safety_mode()`:

```
======================================================================
FAIL: test_tx_hook_on_wrong_safety_mode (test_volkswagen_mqb.TestVolkswagenMqbLongSafety)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jyoung/openpilot/panda/tests/safety/common.py", line 441, in test_tx_hook_on_wrong_safety_mode
    self.assertFalse(self._tx(msg), f"transmit of {addr=:#x} {bus=} from {test_name} was allowed")
AssertionError: 1 is not false : transmit of addr=0x126 bus=0 from VolkswagenMqbStockSafety was allowed

======================================================================
FAIL: test_tx_hook_on_wrong_safety_mode (test_volkswagen_mqb.TestVolkswagenMqbStockSafety)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jyoung/openpilot/panda/tests/safety/common.py", line 441, in test_tx_hook_on_wrong_safety_mode
    self.assertFalse(self._tx(msg), f"transmit of {addr=:#x} {bus=} from {test_name} was allowed")
AssertionError: 1 is not false : transmit of addr=0x126 bus=0 from VolkswagenMqbLongSafety was allowed

----------------------------------------------------------------------
```

I spent some time looking at the code and I don't yet understand why the test is flagging VW MQB for using the same message on different variants of my safety while NOT flagging other ports like Hyundai for the same. I'll keep looking at it, but...

```
======================================================================
FAIL: test_tx_hook_on_wrong_safety_mode (test_honda.TestHondaBoschLongGiraffeSafety)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jyoung/openpilot/panda/tests/safety/common.py", line 441, in test_tx_hook_on_wrong_safety_mode
    self.assertFalse(self._tx(msg), f"transmit of {addr=:#x} {bus=} from {test_name} was allowed")
AssertionError: 1 is not false : transmit of addr=0x30c bus=0 from VolkswagenMqbLongSafety was allowed

======================================================================
FAIL: test_tx_hook_on_wrong_safety_mode (test_honda.TestHondaNidecSafety)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jyoung/openpilot/panda/tests/safety/common.py", line 441, in test_tx_hook_on_wrong_safety_mode
    self.assertFalse(self._tx(msg), f"transmit of {addr=:#x} {bus=} from {test_name} was allowed")
AssertionError: 1 is not false : transmit of addr=0x30c bus=0 from VolkswagenMqbLongSafety was allowed

======================================================================
FAIL: test_tx_hook_on_wrong_safety_mode (test_subaru.TestSubaruSafety)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jyoung/openpilot/panda/tests/safety/common.py", line 441, in test_tx_hook_on_wrong_safety_mode
    self.assertFalse(self._tx(msg), f"transmit of {addr=:#x} {bus=} from {test_name} was allowed")
AssertionError: 1 is not false : transmit of addr=0x122 bus=0 from VolkswagenMqbLongSafety was allowed
```

These look to be legitimate collisions of transmitted messages between car ports. I could add them to the ad-hoc exceptions already written into this test, but I have to ask... if we're going to swiss-cheese the test into "never do this, unless we do, then whatever", what's the point of having this test?

File: board/safety/safety_volkswagen.h [hunk 1]
```diff
@@ -383,6 +445,10 @@ static int volkswagen_fwd_hook(int bus_num, CANPacket_t *to_fwd) {
       if ((addr == volkswagen_torque_msg) || (addr == volkswagen_lane_msg)) {
         // OP takes control of the Heading Control Assist and Lane Departure Warning messages from the camera
         bus_fwd = -1;
+        } else if (volkswagen_longitudinal && ((addr == volkswagen_acc_accel_msg_1) ||
+                  (addr == volkswagen_acc_accel_msg_2) || (addr == volkswagen_acc_hud_msg))) {
```
Comment (adeebshihadeh) [inline board/safety/safety_volkswagen.h:449]: wrong indentation level

File: board/safety/safety_volkswagen.h [hunk 2]
```diff
@@ -171,18 +186,33 @@ static int volkswagen_mqb_rx_hook(CANPacket_t *to_push) {
...
+    if (volkswagen_longitudinal) {
+      if (addr == MSG_GRA_ACC_01) {
+        // Exit controls on Cancel, otherwise, enter controls on Set or Resume
+        // Signal: GRA_ACC_01.GRA_Abbrechen
+        // Signal: GRA_ACC_01.GRA_Tip_Setzen
+        // Signal: GRA_ACC_01.GRA_Tip_Wiederaufnahme
+        if ((GET_BYTE(to_push, 2) & 0x9) != 0) {
+          controls_allowed = 1;
+        }
+        if ((GET_BYTE(to_push, 1) & 0x20) != 0) {
```
Comment (adeebshihadeh) [inline board/safety/safety_volkswagen.h:198]: add a GET_BIT macro so it's easy when checking against the DBC

File: board/safety/safety_volkswagen.h [hunk 3]
```diff
@@ -124,8 +135,12 @@ static const addr_checks* volkswagen_mqb_init(int16_t param) {
 
   controls_allowed = false;
   relay_malfunction_reset();
+  volkswagen_longitudinal = GET_FLAG(param, VOLKSWAGEN_PARAM_LONG);
```
Comment (adeebshihadeh) [inline board/safety/safety_volkswagen.h:138]: get this behind the debug flag for now

**Comment by adeebshihadeh** [comment]
The test is still around with the exception so that were aware of the list of violations and immediately know about any new ones.

**Comment by jyoung8607** [comment]
This safety has been road tested as working, but planning to hold in draft until I have the Front Assist dash warning operating per current guidelines. I'll almost certainly need to have openpilot send ACC_10 and have Panda pass it.

You may or may not care for my edits to `test_tx_hook_on_wrong_safety_mode()`. Let me know if you want it done a different way. I made the rest of the changes you requested.

**Comment by jyoung8607** [comment]
Ready for review.

We still have trouble with a subset of newer cars that briefly throw "Brake!" alerts on disengage. I know what's causing this, it's essentially a reconciliation check between the factory ACC radar and the instrument cluster. I'm working out the best way to silence it without disrupting factory FCW/AEB. It's possible I'll need a follow-up PR to add another message to the filter list.

Test route: `3cfdec54aa035f3f|2021-12-07--14-19-01`

I request that route be cached for both Panda `test_safety_replay` (I tested with segment 0 locally) and openpilot `test_routes` use. I'm not ready to freeze things for an openpilot process_replay route though.

**Comment by jyoung8607** [comment]
Side note: sometime in the future I'm going to submit a PR to split the MQB and PQ safeties into separate files. I'm not really getting the code-sharing I thought I would back in the day, the little sharing I get adds more LoC than it removes, and upcoming PQ long support will make the abstraction even more tortured. The safety tests are already split.

**Comment by jyoung8607** [comment]
Temporarily withdrawing this PR, to be resubmitted based on #830 per my last comment.

**Pull request closed without merging**